dpcpp/base/kernel_launch.dp.hpp

```diff
@@ -0,0 +1,242 @@
+
+
+#ifdef GKO_KERNEL
+#error "Only one kernel_launch.hpp file can be included at a time."
+#else
+#define GKO_KERNEL
+#endif
+
+
+namespace gko {
+namespace kernels {
+namespace dpcpp {
+
+
+template <typename ValueType>
+struct matrix_accessor {
+    ValueType *data;
+    size_type stride;
+
+    ValueType &operator()(size_type row, size_type col)
+    {
+        return data[row * stride + col];
+    }
+
+    ValueType &operator[](size_type idx) { return data[idx]; }
+};
+
+
+template <typename T>
+struct to_device_type_impl {
+    using type = std::decay_t<T>;
+    static type map_to_device(T in) { return in; }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<ValueType>;
+    static type map_to_device(matrix::Dense<ValueType> *mtx)
+    {
+        return {mtx->get_values(), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const matrix::Dense<ValueType> *&> {
+    using type = matrix_accessor<const ValueType>;
+    static type map_to_device(const matrix::Dense<ValueType> *mtx)
+    {
+        return {mtx->get_const_values(), mtx->get_stride()};
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<Array<ValueType> &> {
+    using type = ValueType *;
+    static type map_to_device(Array<ValueType> &array)
+    {
+        return array.get_data();
+    }
+};
+
+template <typename ValueType>
+struct to_device_type_impl<const Array<ValueType> &> {
+    using type = const ValueType *;
+    static type map_to_device(const Array<ValueType> &array)
+    {
+        return array.get_const_data();
+    }
+};
+
+
+template <typename T>
+typename to_device_type_impl<T>::type map_to_device(T &&param)
+{
+    return to_device_type_impl<T>::map_to_device(param);
+}
+
+
+template <typename ValueType>
+struct compact_dense_wrapper {
+    ValueType *data;
+};
+
+
+template <typename ValueType>
+compact_dense_wrapper<ValueType> compact(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {mtx->get_values()};
+}
+
+
+template <typename ValueType>
+compact_dense_wrapper<const ValueType> compact(
+    const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_stride() == mtx->get_size()[1]);
+    return {mtx->get_const_values()};
+}
+
+
+template <typename ValueType>
+ValueType *vector(matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return mtx->get_values();
+}
+
+
+template <typename ValueType>
+const ValueType *vector(const matrix::Dense<ValueType> *mtx)
+{
+    GKO_ASSERT(mtx->get_size()[0] == 1 ||
+               (mtx->get_size()[1] == 1 && mtx->get_stride() == 1));
+    return mtx->get_const_values();
+}
+
+
+template <typename T>
+struct device_unpack_1d_impl {
+    using type = T;
+    static type unpack(T param, size_type, size_type) { return param; }
```

Comment:
Again, I don't get the reason for all these unused parameters in this function and the next.